### PR TITLE
bwl: nl80211_client: add set_tx_power_limit API

### LINKED ISSUE #983: [TASK] set_tx_power support
## Description

Add support for setting tx power limit in the agent per radio. This is not available in hostapd, so we need to use nl80211 for that.

### Steps

- [x] Add a new method to nl80211_client to set tx power limit using NL80211_CMD_SET_WIPHY.
- [x] Add a nl80211_client to the base_wlan_hal which will be used by all BWL backends to get and set WLAN nl80211 generic commands.
- [x] Move `set_tx_power_limit` API to ap_wlan_hal
- [x] Implement the stub in `set_tx_power_limit` to use the `nl80211_client` for setting tx_power limit
- [ ] Implement set_tx_power limit in dummy BWL implementation which will store the limit
- [ ] Enhance channel selection dummy test to set new tx power limit as done in the certification

Edit - the last two are a bit tricky with the current implementation since the `nl80211_client` doesn't have access to the `m_radio_info` which holds the tx power reading - this is read from hostapd using `radio_info` command which reports the tx power via the TxPower parameter, and is saved in the `m_radio_info` structure which belongs to the `base_wlan_hal` class.
So, unless there are better ideas I suggest not to implement this in dummy mode at all. We will have coverage for this in real devices test flows.

### PR DESCRIPTION
**Pull request by tomereli**
Some EasyMesh flows require setting tx power limit, which is not
supported via hostapd (tx_power is readable via radio_info command),
Add set_tx_power_limit() API to the nl80211 client implementation which
uses NL80211_CMD_SET_WIPHY to set the limit.

Currently doesn't work since set tx power is not supported by the driver, but this can be merged since we get the same status as when doing this with `iw` - so marking as ready for review.
**Tested on AXEPOINT3000 (UGW) and everything works as expected.**

Since this PR changes some channel selection handling, triggered gitlan CI pipeline to verify nothing broke, which tests the following tests and verifies autoconfig, onboarding and channel selection:
```
MAP-4.2.1
MAP-4.5.2_ETH_FH5GL
MAP-4.5.3_ETH_FH5GL
MAP-4.5.3_ETH_FH5GH
```
https://gitlab.com/prpl-foundation/prplMesh/pipelines/128634565

Fixes #983
 
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

### CODE REVIEW

File: common/beerocks/bwl/dummy/nl80211_client_dummy.h
Comment (mariomaz) [inline common/beerocks/bwl/dummy/nl80211_client_dummy.h]: I think it can be a physical radio interface as well as a virtual interface

File: common/beerocks/bwl/include/bwl/ap_wlan_hal.h
Comment (mariomaz) [inline common/beerocks/bwl/include/bwl/ap_wlan_hal.h]: I think parameter should be uint32_t

File: common/beerocks/bwl/include/bwl/nl80211_client.h
Comment (mariomaz) [inline common/beerocks/bwl/include/bwl/nl80211_client.h]: I think it works both with radio or virtual interface name

File: common/beerocks/bwl/shared/nl80211_client_impl.cpp
Comment (mariomaz) [inline common/beerocks/bwl/shared/nl80211_client_impl.cpp]: There must be some kind of return code to check if operation succeeded or not

File: common/beerocks/bwl/shared/nl80211_client_impl.h
Comment (mariomaz) [inline common/beerocks/bwl/shared/nl80211_client_impl.h]: Radio interface

File: common/beerocks/bwl/shared/nl80211_client_impl.h
Comment (mariomaz) [inline common/beerocks/bwl/shared/nl80211_client_impl.h]: What do you mean by 'object'?

File: common/beerocks/bwl/common/base_wlan_hal.cpp
Comment (mariomaz) [inline common/beerocks/bwl/common/base_wlan_hal.cpp]: nl80211_client is a dependency and thus should be injected, not hard-coded.
If we ever create a unit test for this class, nl80211_client should be provided as a parameter.

**Review by mariomaz — APPROVED** [review]
Only nitpick comments so approving.

File: common/beerocks/bwl/shared/nl80211_client_impl.h
Comment (tomereli) [reply common/beerocks/bwl/shared/nl80211_client_impl.h]: typo

File: common/beerocks/bwl/common/base_wlan_hal.cpp
Comment (tomereli) [reply common/beerocks/bwl/common/base_wlan_hal.cpp]: I'm not sure I follow - can you explain what is hard coded here and what do you mean by injecting it?

Comment (mariomaz) [reply common/beerocks/bwl/common/base_wlan_hal.cpp]: `nl80211_client` is a dependency for `base_wlan_hal`. `base_wlan_hal` should not create it. `nl80211_client` should be injected as a parameter to the constructor of `base_wlan_hal`.

But this is only actually required if we want to create a unit test for `base_wlan_hal`. In such unit test, we would pass a mock `nl80211_client` instead of the real implementation.

**Comment by mariomaz** [comment]
BTW, I've realized just now that you've moved `set_tx_power_limit` from `ap_wlan_hal_*` to `base_wlan_hal`. This is a base class for `ap_wlan_hal_*`, `base_wlan_hal_*`, `mon_wlan_hal_*` and `sta_wlan_hal_*`classes so `set_tx_power_limit` will be available to all of them. Was that the intention or a side effect? I think there will be more instances of `nl80211_client` that what they are really needed.

Moreover, since `ap_wlan_hal_*` classes extend both `base_wlan_hal` and `base_wlan_hal_*` there will be two instances of `nl80211_client`. Again, more than needed. 

I think `nl80211_client` should be a member of `ap_wlan_hal_*` and not of `base_wlan_hal`

**Comment by tomereli** [comment]
> BTW, I've realized just now that you've moved `set_tx_power_limit` from `ap_wlan_hal_*` to `base_wlan_hal`. This is a base class for `ap_wlan_hal_*`, `base_wlan_hal_*`, `mon_wlan_hal_*` and `sta_wlan_hal_*`classes so `set_tx_power_limit` will be available to all of them. Was that the intention or a side effect? I think there will be more instances of `nl80211_client` that what they are really needed.
> 
> Moreover, since `ap_wlan_hal_*` classes extend both `base_wlan_hal` and `base_wlan_hal_*` there will be two instances of `nl80211_client`. Again, more than needed.
> 
> I think `nl80211_client` should be a member of `ap_wlan_hal_*` and not of `base_wlan_hal`

True, didn't think about this. Thanks. I added the don't merge tag for now.

File: common/beerocks/bwl/shared/nl80211_client_impl.cpp
Comment (tomereli) [reply common/beerocks/bwl/shared/nl80211_client_impl.cpp]: I couldn't find this - from what I could understand, there is a default callback for error so I don't need to handle errors here.

File: common/beerocks/bwl/dummy/nl80211_client_dummy.h
Comment (tomereli) [reply common/beerocks/bwl/dummy/nl80211_client_dummy.h]: I think it will cause confusion, so lets keep it for radio only.

File: common/beerocks/bwl/include/bwl/nl80211_client.h
Comment (tomereli) [reply common/beerocks/bwl/include/bwl/nl80211_client.h]: I checked on UGW where I have only radio and it works, I didn't check VAP so I'll change the comment to say only the radio name is supported.

File: common/beerocks/bwl/common/base_wlan_hal.cpp
Comment (tomereli) [reply common/beerocks/bwl/common/base_wlan_hal.cpp]: Done

File: common/beerocks/bwl/include/bwl/ap_wlan_hal.h
Comment (tomereli) [reply common/beerocks/bwl/include/bwl/ap_wlan_hal.h]: This was already like that apparently and I reverted this original change so I prefer not to change this as part of this PR.

**Comment by tomereli** [comment]
@mariomaz I refactored this PR based on your last comment, so I think this requires a new review.

File: common/beerocks/bwl/include/bwl/nl80211_client.h
Comment (arnout) [inline common/beerocks/bwl/include/bwl/nl80211_client.h]: Didn't you say that "object" was a typo and that you fixed it?

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp
Comment (arnout) [inline common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp]: I believe the std::move is redundant here because unique_ptr only has a move constructor, no copy constructor. (nitpick)

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp
Comment (arnout) [inline common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp]: @mariomaz I still don't really understand how the mocking would work. Now the nl80211_client_factory is hardcoded here, in ap_wlan_hal_create. If we would want to create a unit test for ap_wlan_hal_dwpal, we'd have to create a "mocking" factory function to replace this one, correct?

File: common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp
Comment (arnout) [inline common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp]: IMO this could have been squashed into the first commit.

File: agent/src/beerocks/slave/ap_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/ap_manager_thread.cpp:584]: Since nothing has changed, refreshing radio info shouldn't be needed.

Looking at it in more detail, it's even worse. Setting the TX power will not be reflected in the result of get_tx_power(), because that is using the tx power limit from the supported channels, and that, in turn, is purely based on the regulatory limits and the hardware limit, *not* on the power limit set by NL80211_ATTR_WIPHY_TX_POWER_SETTING. So I think we need to add an actual power limit setting in bwl::RadioInfo and make sure it gets filled in by refresh_radio_info(). But that's for another PR...

**Review by arnout — APPROVED** [review]
Just one remaining typo in a comment.

File: agent/src/beerocks/slave/ap_manager_thread.cpp
Comment (tomereli) [reply agent/src/beerocks/slave/ap_manager_thread.cpp:584]: It looks like @mariomaz already included the max tx power using `NL80211_FREQUENCY_ATTR_MAX_TX_POWER` in #977 - I think we should decide how we want to proceed with @mariomaz 's PR before we merge this one.
I have UGW so I can test it correctly all the way, no need to rush and merge this half way done.

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp
Comment (tomereli) [reply common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp]: Apparently it is not :/

File: agent/src/beerocks/slave/ap_manager_thread.cpp
Comment (tomereli) [reply agent/src/beerocks/slave/ap_manager_thread.cpp:584]: So apparently my previous comment is wrong, we don't need to rebase on #977 in order to read the tx power since it is reflected in `TxPower` when running `hostapd_cli -iwlan0 radio_info` hence available in `m_radio_info.conducted_power`. Added a commit which fixes this.

File: common/beerocks/bwl/shared/nl80211_client_impl.cpp
Comment (vitalybu) [inline common/beerocks/bwl/shared/nl80211_client_impl.cpp]: I think that you should either validate that `local_interface_name` is not empty, or at least add `''` to the log (so that it'll be easier to spot an empty value):
```suggestion
        LOG(ERROR) << "Failed to read the index of interface '" << local_interface_name << "': "
```

File: common/beerocks/bwl/shared/nl80211_client_impl.cpp
Comment (vitalybu) [inline common/beerocks/bwl/shared/nl80211_client_impl.cpp:206]: Shouldn't we allow returning the power setting to `NL80211_TX_POWER_AUTOMATIC`?
Or perhaps even support all TX_POWER modes:
- NL80211_TX_POWER_AUTOMATIC
- NL80211_TX_POWER_FIXED
- NL80211_TX_POWER_LIMITED

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp
Comment (vitalybu) [inline common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:1593]: Since now `nl80211_client` is passed to the constructor (and has a default value of `nullptr`), you can end up with a null reference exception here.
You should either remove the default value and make sure it's not `nullptr`, or check if it's not `nullptr` before every call.

File: common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
Comment (vitalybu) [inline common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:575]: dwpal (previous commit) comments apply here as well

File: common/beerocks/bwl/shared/nl80211_client_impl.cpp
Comment (tomereli) [reply common/beerocks/bwl/shared/nl80211_client_impl.cpp:206]: The only thing which is needed by EasyMesh is setting power limit.

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp
Comment (tomereli) [reply common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:1593]: But in dwpal we always pass it, and override the default value - and if it fails then we assert with LOG_IF(FATAL) - do you still think it is necessary to check every call?

Comment (vitalybu) [reply common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:1593]: No, I think you should remove the default value from the constructor and force passing a valid instance, while also validating that it's not nullptr.

File: common/beerocks/bwl/shared/nl80211_client_impl.cpp
Comment (vitalybu) [reply common/beerocks/bwl/shared/nl80211_client_impl.cpp:206]: ¯\\_(ツ)_/¯

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp
Comment (tomereli) [reply common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:1593]: OK, done. I removed the default value and just moved the initialization to `base_wlan_hal_*`. IMO it makes much more sense there instead of manually specifying it for every derived class.

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp
Comment (tomereli) [reply common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp]: I moved the `create_instance` to `base_wlan_hal_XXX` constructor since the nl80211_client can be and will be used by the base class to update radio info which is implemented in the base class.

As @vitalybu said:
>Dependency injection is useful when you want to "inject" different implementations of the same interface to some class. I'm just not sure if it's the case here. Why would we ever use a non-DWPAL nl80211 client with DWPAL BWL?

File: common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp
Comment (tomereli) [reply common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp]: IMHO it makes more sense as a separate commit.

File: common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.h
Comment (arnout) [inline common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.h:92]: What is the point of moving it to base_wlan_hal_dwpal? I though we said that it wouldn't be used by the monitor or station HALs...

More importantly though: if you move it here, it should also be taken as an argument in the base_wlan_hal_dwpal constructor and set there, not in the derived class's constructor.

**Review by arnout — APPROVED** [review]
Still looks good to me, with one minor comment.

File: common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.h
Comment (tomereli) [reply common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.h:92]: The point is that the `refresh_radio_info` is implemented in the base, so I thought that in order to use @mariomaz's #977 to read the TxPower directly from nl80211 I will need to do that from `refresh_radio_info`.
It turns out that the `TxPower` we get from hostapd is fine, but nevertheless we will probably update `refresh_radio_info` to complement the radio info parameters we get from hostapd with the values we read from nl80211, so I decided to keep it there.

Comment (tomereli) [reply common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.h:92]: >More importantly though: if you move it here, it should also be taken as an argument in the base_wlan_hal_dwpal constructor and set there, not in the derived class's constructor.

What's wrong with instantiating it in `base_wlan_hal_dwpal` like I did here? It is not set in the derived class constructor:
```
base_wlan_hal_dwpal::base_wlan_hal_dwpal(HALType type, std::string iface_name,
                                         hal_event_cb_t callback, hal_conf_t hal_conf)
    : base_wlan_hal(type, iface_name, IfaceType::Intel, callback, hal_conf),
      beerocks::beerocks_fsm<dwpal_fsm_state, dwpal_fsm_event>(dwpal_fsm_state::Delay),
      m_nl80211_client(nl80211_client_factory::create_instance())
{
    LOG_IF(!m_nl80211_client, FATAL) << "Failed to create nl80211_client instance";
    // Initialize the FSM
    fsm_setup();
}
```

Comment (tomereli) [reply common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.h]:
```
base_wlan_hal_dwpal::base_wlan_hal_dwpal(HALType type, std::string iface_name,
                                         hal_event_cb_t callback, hal_conf_t hal_conf)
    : base_wlan_hal(type, iface_name, IfaceType::Intel, callback, hal_conf),
      beerocks::beerocks_fsm<dwpal_fsm_state, dwpal_fsm_event>(dwpal_fsm_state::Delay),
      m_nl80211_client(nl80211_client_factory::create_instance())
{
    LOG_IF(!m_nl80211_client, FATAL) << "Failed to create nl80211_client instance";
    // Initialize the FSM
    fsm_setup();
}
```

Comment (arnout) [reply common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.h:92]: I'm sorry, I misread the code.

**Pull request merged**
